### PR TITLE
 Add support for regional STS endpoints

### PR DESCRIPTION
**Pull request by stanhu**
This pull request is based off https://github.com/fog/fog-aws/pull/605/files and cleans up and fixes the implementation and test cases.

### CODE REVIEW

**Comment by stanhu** [comment]
@geemus Could you take a look at this?

File: lib/fog/aws/credential_fetcher.rb [hunk 1]
```diff
@@ -23,7 +21,7 @@ def fetch_credentials(options)
           if options[:use_iam_profile]
             begin
               role_data = nil
-              region = options[:region]
+              region = options[:region] || ENV["AWS_DEFAULT_REGION"]
```
Comment (stanhu) [inline lib/fog/aws/credential_fetcher.rb:24]: I think `AWS_REGION` might have been used before, but I think `AWS_DEFAULT_REGION` is the new environment variable: https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html.

File: lib/fog/aws/credential_fetcher.rb [hunk 2]
```diff
@@ -44,7 +42,15 @@ def fetch_credentials(options)
                   :WebIdentityToken => File.read(options[:aws_web_identity_token_file] || ENV.fetch("AWS_WEB_IDENTITY_TOKEN_FILE")),
                   :Version => "2011-06-15",
                 }
-                connection = options[:connection] || Excon.new(STS_GLOBAL_ENDPOINT, :query => params)
+
+                sts_endpoint =
+                  if ENV["AWS_STS_REGIONAL_ENDPOINTS"] == "regional"
```
Comment (stanhu) [inline lib/fog/aws/credential_fetcher.rb:47]: Maybe we need to check `region` as well.

File: lib/fog/aws/credential_fetcher.rb [hunk 3]
```diff
@@ -44,7 +42,15 @@ def fetch_credentials(options)
                   :WebIdentityToken => File.read(options[:aws_web_identity_token_file] || ENV.fetch("AWS_WEB_IDENTITY_TOKEN_FILE")),
                   :Version => "2011-06-15",
                 }
-                connection = options[:connection] || Excon.new(STS_GLOBAL_ENDPOINT, :query => params)
+
+                sts_endpoint =
+                  if ENV["AWS_STS_REGIONAL_ENDPOINTS"] == "regional" && region
```
Comment (geemus) [inline lib/fog/aws/credential_fetcher.rb:47]: Part of this still seems a little weird to me. I guess it feels like maybe the key should just be something like `true`? ie `ENV["AWS_STS_REGIONAL_ENDPOINTS"]` is binary, so making it that explicitly seems a bit cleaner to me than using a special `regional` key. Does that make sense?

Comment (geemus) [reply lib/fog/aws/credential_fetcher.rb:47]: Ah, wait, I see. It matches the keys/values from the AWS CLI. That makes more sense.

**Review by geemus — APPROVED** [review]
LGTM, thanks for helping finish this up!

**Comment by stanhu** [comment]
@geemus Thanks for the quick review! Would you mind tagging a new release?

**Comment by MichaelGerhart** [comment]
@stanhu thanks for getting this finished up! I meant to come back to it but there’s always something…

**Comment by geemus** [comment]
@MichaelGerhart no worries, we've all been there.

**Comment by geemus** [comment]
@stanhu is Monday soon enough? I'm always wary of releasing right before the weekend (when I won't be checking email and stuff as much).

**Comment by stanhu** [comment]
Sure.

**Comment by geemus** [comment]
Awesome, thanks!

**Comment by geemus** [comment]
Released in v3.12.0.

**Pull request merged**